### PR TITLE
Kafka Producer plugin

### PR DESCRIPTION
**Pull request by hir12111**
- Implementation of a plugin that allows sending frames to one or more Kafka servers asynchronously.
- Unit tests
- Kafka cmake file
- Travis dependency installer

A few decisions were made in this implementation:

 - librdkafka C library was used instead of the C++ version as it is more flexible.
 - The message header is json (for more flexibility and easy to parse), the message structure is: [2 bytes header size] + [json header] + [frame data]
 - The header contains the following keys: "data_size", "data_type", "frame_number", "dims", "frame_offset", "acquisition_id", "compression" and optionally "parameters" (object containing frame parameters)
 - The plugin is optional
 - The message is sent without blocking (or waiting for ack)
 - Default partition is automatic partitioning (RD_KAFKA_PARTITION_UA)
 - It allows choosing which dataset will be published, defaults to "data"
 - It allows choosing in which topic will be published, defaults to "data"
 - If many topics or datasets are required, the plugin could be instantiated many times with different configuration.

I've already tested it with Tom in our kafka cluster and everything works fine.
EDIT: it has also been tested with a real Eiger detector

### CODE REVIEW

**Comment by ulrikpedersen** [comment]
Thanks @hir12111! 

We want to test this with at least one of our detector systems (preferably Eiger) before merging this PR.

**Comment by hir12111** [comment]
I was using Stroustrup code style in my IDE, unfortunately there isn't any predefined code style that matches this project, so I fixed it manually in the last commit

**Comment by hir12111** [comment]
I tested it with the Eiger simulator and it worked fine, then, Matt and I went to VMXi and tested it with a real Eiger, the test worked and the Kafka server got the data without problems. 
In future we might try doing load testing,  by pushing as many frames as possible (once SC gets the servers ready)

File: frameProcessor/src/KafkaProducerPlugin.cpp [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    // Make sure statistics are updated
+    poll_delivery_message_report_queue();
+
+    reply.set_param(get_name() + "/" + KafkaProducerPlugin::CONFIG_SENT,
+                    frames_sent_);
+    reply.set_param(get_name() + "/" + KafkaProducerPlugin::CONFIG_LOST,
+                    frames_lost_);
+    reply.set_param(get_name() + "/" + KafkaProducerPlugin::CONFIG_ACK,
+                    frames_ack_);
+  }
```
Comment (GDYendell) [inline frameProcessor/src/KafkaProducerPlugin.cpp]: The above config items should be added to the readback here.

**Comment by hir12111** [comment]
I've fixed what Gary mentioned in the last commit, now all the configuration parameters are included in requestConfiguration and the stats are in status.
I have also included a new parameter for configuring the topic name (independently of the dataset name), it defaults to "data" if not set.

**Comment by MattTaylorDLS** [comment]
Did we find out if these are the things that the intended kafka consumers are going to want? Is there a specification drawn up somewhere?

**Comment by ulrikpedersen** [comment]
> Did we find out if these are the things that the intended kafka consumers are going to want? Is there a specification drawn up somewhere?

Not yet. That's a bit of a bee's nest to dive into and I need to clear some of my own backlog before I can properly dedicate some time to this...

File: frameProcessor/include/KafkaProducerPlugin.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
+//
+// Created by hir12111 on 25/03/19.
```
Comment (skyfrench) [inline frameProcessor/include/KafkaProducerPlugin.h]: Replace fed ID with author name

File: frameProcessor/src/KafkaProducerPluginLib.cpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
+//
+// Created by hir12111 on 25/03/19.
```
Comment (skyfrench) [inline frameProcessor/src/KafkaProducerPluginLib.cpp]: Fed ID -> name

File: frameProcessor/src/KafkaProducerPlugin.cpp [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+#include "rapidjson/writer.h"
+#include "rapidjson/stringbuffer.h"
+
+#include "KafkaProducerPlugin.h"
+#include "version.h"
+
+namespace FrameProcessor {
+
+  // Callback function used to report status of a delivered message
+  static void kafka_message_callback(rd_kafka_t *kafka_producer,
```
Comment (skyfrench) [inline frameProcessor/src/KafkaProducerPlugin.cpp]: For all functions should there be more detailed comments? Detailed comment specifying parameters and return in source code, and single line / trailing comment in header for methods?

**Pull request merged**